src/framework/jobState/FileSystemJobState/indices.ts

```diff
@@ -0,0 +1,89 @@
+import { Entity, Relationship } from '../../types';
+import { GraphObjectIteratee } from '../types';
+
+import {
+  walkDirectory,
+  WalkDirectoryIterateeInput,
+} from '../../../cacheDirectory';
+
+import { buildIndexDirectoryPath } from './path';
+
+interface IterateIndexInput<GraphObject> {
+  cacheDirectory?: string;
+  type: string;
+  iteratee: GraphObjectIteratee<GraphObject>;
+}
+
+export async function iterateEntityTypeIndex({
+  cacheDirectory,
+  type,
+  iteratee,
+}: IterateIndexInput<Entity>) {
+  const path = buildIndexDirectoryPath({
+    collectionType: 'entities',
+    type,
+  });
+
+  await walkDirectory({
+    cacheDirectory,
+    path,
+    iteratee: async (input) => {
+      const object = parseData(input);
+      if (isObjectFlushedEntityData(object)) {
+        for (const entity of object.entities) {
+          await iteratee(entity);
+        }
+      }
+    },
+  });
+}
+
+export async function iterateRelationshipTypeIndex({
+  cacheDirectory,
+  type,
+  iteratee,
+}: IterateIndexInput<Relationship>) {
+  const path = buildIndexDirectoryPath({
+    collectionType: 'relationships',
+    type,
+  });
+
+  await walkDirectory({
+    cacheDirectory,
+    path,
+    iteratee: async (input) => {
+      const object = parseData(input);
+      if (isObjectFlushedRelationshipData(object)) {
+        for (const relationship of object.relationships) {
+          await iteratee(relationship);
+        }
+      }
+    },
+  });
+}
```

Comment:
Definitely seems like `iterateEntityTypeIndex` and `iterateRelationshipTypeIndex` could have their implementation combined into a third function that both wrap, but that may not be worth it to you.